portforward/forwarder.go

```diff
@@ -0,0 +1,272 @@
+				if err != nil {
+					logger.Error(ctx, "failed to dial remote",
+						slog.F("dial_address", dialAddress),
+						slog.Error(err))
+					return
+				}
+				defer remoteConn.Close()
+				logger.Debug(ctx, "dialed remote",
+					slog.F("remote_addr", netConn.RemoteAddr()))
+
+				agentssh.Bicopy(ctx, netConn, remoteConn)
+				logger.Debug(ctx, "connection closing",
+					slog.F("remote_addr", netConn.RemoteAddr()))
+			}(netConn)
+		}
+	}()
+
+	return nil
+}
+
+func (f *LocalForwarder) Stop() error {
+	if !f.active.Load() {
+		return nil
+	}
+
+	if f.cancel != nil {
+		f.cancel()
+	}
+	if f.listener != nil {
+		_ = f.listener.Close()
+	}
+	f.wg.Wait()
+	return nil
+}
+
+func (f *LocalForwarder) IsActive() bool {
+	return f.active.Load()
+}
+
+func (f *LocalForwarder) Spec() Spec {
+	return f.spec
+}
+
+// manager implements the Manager interface.
+type manager struct {
+	forwarders map[string]Forwarder
+	opts       Options
+	mu         sync.RWMutex
+}
+
+// NewManager creates a new port forwarding manager.
+func NewManager(opts Options) Manager {
+	return &manager{
+		forwarders: make(map[string]Forwarder),
+		opts:       opts,
+	}
+}
+
+func (m *manager) Add(spec Spec) (Forwarder, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	key := fmt.Sprintf("%s:%s:%d", spec.Network, spec.ListenHost, spec.ListenPort)
+	if _, exists := m.forwarders[key]; exists {
+		return nil, xerrors.Errorf("forwarder already exists for %s", key)
+	}
+
+	// Test if we can actually bind to the port before adding the forwarder
+	listenAddress := netip.AddrPortFrom(spec.ListenHost, spec.ListenPort)
+	testListener, err := m.opts.Listener.Listen(spec.Network, listenAddress.String())
+	if err != nil {
+		return nil, xerrors.Errorf("cannot bind to '%s://%s': %w", spec.Network, listenAddress.String(), err)
+	}
+	// Close the test listener immediately since we just wanted to verify we can bind
+	_ = testListener.Close()
+
+	forwarder := NewLocal(spec, m.opts)
+	m.forwarders[key] = forwarder
+	return forwarder, nil
+}
+
+func (m *manager) Remove(spec Spec) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	key := fmt.Sprintf("%s:%s:%d", spec.Network, spec.ListenHost, spec.ListenPort)
+	forwarder, exists := m.forwarders[key]
+	if !exists {
+		return xerrors.Errorf("forwarder not found for %s", key)
+	}
+
+	err := forwarder.Stop()
+	delete(m.forwarders, key)
+	return err
+}
+
+func (m *manager) List() []Forwarder {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	forwarders := make([]Forwarder, 0, len(m.forwarders))
+	for _, f := range m.forwarders {
+		forwarders = append(forwarders, f)
+	}
+	return forwarders
+}
+
+func (m *manager) Start(ctx context.Context) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	for _, forwarder := range m.forwarders {
+		if !forwarder.IsActive() {
+			if err := forwarder.Start(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (m *manager) Stop() error {
```

Comment:
```suggestion
func (m *manager) Close() error {
```

so it implements `io.Closer`